### PR TITLE
[RLlib] Better utils for flattening complex inputs and enable prev-actions for LSTM/attention for complex action spaces.

### PR DESCRIPTION
**Pull request by sven1977**
This PR:
- Improves tf/torch/np utils for flattening complex inputs (into a single 1D tensor).
- Enables prev-actions for LSTM/attention for complex action spaces. `lstm_use_prev_action` and `attention_use_n_prev_actions`  would only work for simple Box or Discrete action spaces.
- Adds two new test cases for the above capabilities (LSTM and attention nets w/ complex action spaces and prev-a enabled).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/rnn_sequencing.py [hunk 1]
```diff
@@ -176,8 +176,16 @@ def add_time_dimension(padded_inputs: TensorType,
...
-            [new_batch_size, max_seq_len] + list(padded_inputs.shape[1:]))
+        new_shape = tf.squeeze(
+            tf.stack(
+                [
+                    tf.expand_dims(new_batch_size, axis=0),
+                    tf.expand_dims(max_seq_len, axis=0),
+                    tf.shape(padded_inputs)[1:]
+                ],
+                axis=0))
+        #    (
```
Comment (gjoliver) [inline rllib/policy/rnn_sequencing.py:187]: what are these comments? should remove?

File: rllib/utils/numpy.py [hunk 2]
```diff
@@ -270,9 +272,87 @@ def lstm(x,
...
+
+    flat_inputs = tree.flatten(inputs)
+    flat_spaces = tree.flatten(spaces_struct) if spaces_struct is not None \
+        else [None] * len(flat_inputs)
+
+    out = []
+    for input_, space in zip(flat_inputs, flat_spaces):
+        assert isinstance(input_, np.ndarray)
+        # One-hot encoding.
+        if isinstance(space, (Discrete, MultiDiscrete)):
```
Comment (gjoliver) [inline rllib/utils/numpy.py:334]: a minor thing, this if statement seems unnecessary. we are checking individual types anyways.
so simply:

if isinstance(space, Discrete):
  ...
elif isinstance(space, MultiDiscrete):
  ...
else:
  ...

File: rllib/utils/numpy.py [hunk 3]
```diff
@@ -270,9 +272,87 @@ def lstm(x,
...
+                    np.concatenate(
+                        [
+                            one_hot(input_[:, i], depth=n).astype(np.float32)
+                            for i, n in enumerate(space.nvec)
+                        ],
+                        axis=-1))
+        # Flatten.
+        else:
+            out.append(
+                input_.reshape([input_.shape[0], -1]).astype(np.float32))
```
Comment (gjoliver) [inline rllib/utils/numpy.py:348]: minor nit: maybe define B = input_.shape[0], then do

input_.reshape([B, -1]).astype(np.float32)

I found it pretty useful to understand the intended shape here, and consistent with the other util funcs.

File: rllib/utils/tf_utils.py [hunk 4]
```diff
@@ -285,6 +285,94 @@ def _create_placeholders(path, value):
...
+        # Store batch and (if applicable) time dimension.
+        if B is None:
+            B = shape[0]
+            if time_axis:
+                T = shape[1]
+
+        # One-hot encoding.
+        if isinstance(space, (Discrete, MultiDiscrete)):
+            if time_axis:
+                input_ = tf.reshape(input_, [B * T])
```
Comment (gjoliver) [inline rllib/utils/tf_utils.py:358]: we don't need to do [B * T, -1] here?
I guess I am just a bit not clear. any chance you can add a test for time_axis=True case?
appreciate.

**Review by gjoliver — COMMENTED** [review]
looks pretty cool actually!
thanks.

File: rllib/utils/numpy.py (hunk 2 above)
Comment (sven1977) [reply rllib/utils/numpy.py:334]: :+1:

File: rllib/utils/tf_utils.py (hunk 4 above)
Comment (sven1977) [reply rllib/utils/tf_utils.py:358]: Hmm, for the Discrete case no (simple ints), but for the MultiDiscrete case, this is probably wrong. Thanks for catching this!

Comment (sven1977) [reply rllib/utils/tf_utils.py:358]: Yeah, I'll add a test for time-axis=True.

File: rllib/policy/rnn_sequencing.py (hunk 1 above)
Comment (sven1977) [reply rllib/policy/rnn_sequencing.py:187]: done

File: rllib/models/tf/attention_net.py [hunk 5]
```diff
@@ -445,33 +452,48 @@ def forward(self, input_dict: Dict[str, TensorType],
...
+                # Fold time-axis into flattened data.
+                flat = tf.reshape(flat, [tf.shape(flat)[0], -1])
+                prev_a_r.append(flat)
             else:
-                prev_a_r.append(
-                    tf.reshape(
-                        tf.cast(input_dict[SampleBatch.PREV_ACTIONS],
-                                tf.float32),
-                        [-1, self.use_n_prev_actions * self.action_dim]))
+                if isinstance(self.action_space, Discrete):
```
Comment (gjoliver) [inline rllib/models/tf/attention_net.py:471]: can you say a few words about this branch? above you have a comment saying "this only properly works ... iff action flattening is disabled".
but this branch is for the case when it's not disabled, so what are we doing here, and what should users expect?
thanks.

File: rllib/models/tf/recurrent_net.py [hunk 6]
```diff
@@ -203,22 +210,35 @@ def forward(self, input_dict: Dict[str, TensorType],
...
-                tf.reshape(tf.cast(prev_a, tf.float32), [-1, self.action_dim]))
+            if self.model_config["_disable_action_flattening"]:
+                prev_a_r.append(
+                    flatten_inputs_to_1d_tensor(
+                        prev_a,
+                        spaces_struct=self.action_space_struct,
+                        time_axis=False,
+                    ))
+            else:
+                if isinstance(self.action_space, (Discrete, MultiDiscrete)):
```
Comment (gjoliver) [inline rllib/models/tf/recurrent_net.py:226]: can you also commnet here :)
we only support Discrete and MultiDiscrete here, everything else is "complex action spaces" that you mentioned above right?

File: rllib/models/torch/attention_net.py [hunk 7]
```diff
@@ -343,32 +350,49 @@ def forward(self, input_dict: Dict[str, TensorType],
...
+                )
+                # Fold time-axis into flattened data.
+                flat = torch.reshape(flat, [flat.shape[0], -1])
+                prev_a_r.append(flat)
             else:
-                prev_a_r.append(
-                    torch.reshape(
-                        input_dict[SampleBatch.PREV_ACTIONS].float(),
-                        [-1, self.use_n_prev_actions * self.action_dim]))
+                if isinstance(self.action_space, Discrete):
```
Comment (gjoliver) [inline rllib/models/torch/attention_net.py:369]: same here actually, thanks.

File: rllib/models/torch/recurrent_net.py [hunk 8]
```diff
@@ -188,22 +195,33 @@ def forward(self, input_dict: Dict[str, TensorType],
...
+            if self.model_config["_disable_action_flattening"]:
+                prev_a_r.append(
+                    flatten_inputs_to_1d_tensor(
+                        prev_a,
+                        spaces_struct=self.action_space_struct,
+                        time_axis=False))
             else:
-                prev_a = input_dict[SampleBatch.PREV_ACTIONS].float()
-            prev_a_r.append(torch.reshape(prev_a, [-1, self.action_dim]))
+                if isinstance(self.action_space, (Discrete, MultiDiscrete)):
```
Comment (gjoliver) [inline rllib/models/torch/recurrent_net.py:210]: and here.

**Review by gjoliver — APPROVED** [review]
only a few comment requests. thanks again for adding the test :)

File: rllib/models/tf/attention_net.py (hunk 5 above)
Comment (sven1977) [reply rllib/models/tf/attention_net.py:471]: Yeah, agree, the comments here are not clear enough. I'll improve these. ...

The point is:
- If action flattening is off (actions come into the model as they were previously sent to the env, either as complex actions or simple discrete, boxes, etc.., doesn't matter), then we can push them through the new flatten util.
- If action flattening is on (soon to be deprecated), we'll behave as-is.

**Pull request merged**